### PR TITLE
Integrate NodeUpdate controller with csi-driver

### LINKED ISSUE #149: Integrate node-update-controller with csi-driver
**Is your feature request related to a problem?/Why is this needed**
- Integrate node-update-controller with csi-driver
- sync of driver and controller will ensure pre-configuration of nodes needed before driver deployment

/feature

**Describe the solution you'd like in detail**
- Deploy node-update-controller as a container inside pod as part of csi-driver deployment

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by Prajyot-Parab**
Signed-off-by: Prajyot-Parab <prajyot.parab2@ibm.com>

**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
- Phase 1 : Integrate node-update-controller with csi-driver

**Which issue(s) this PR fixes**:

Fixes #149

**Special notes for your reviewer**:

**Release note**:
```
Integrate NodeUpdate controller with csi-driver
```

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -35,17 +36,23 @@ bin:
...
+	CGO_ENABLED=0 go build -ldflags ${LDFLAGS} -o bin/node-update-controller ./adhoc-controllers/
+
 .PHONY: test
 test:
 	go test -v -race ./cmd/... ./pkg/...
 
 .PHONY: image
 image:
 	docker build -t $(REGISTRY)/$(IMG):$(TAG) . --target debian-base
+	docker build -t $(REGISTRY)/$(IMG_NODE_UPDATE_CONTROLLER):$(TAG) . --target debian-base-node-update-controller
```
Comment (Madhan-SWE) [inline Makefile:50]: No need to build node-update-controller every time when csi driver is built.
Create separate target for building node-update-controller image.

File: Makefile [hunk 2]
```diff
@@ -35,17 +36,23 @@ bin:
...
 
 .PHONY: push
 push:
 	docker push $(REGISTRY)/$(IMG):$(TAG)
+	docker push $(REGISTRY)/$(IMG_NODE_UPDATE_CONTROLLER):$(TAG)
```
Comment (Madhan-SWE) [inline Makefile:55]: No need to build node-update-controller every time when csi driver is built.
Create separate target for pushing node-update-controller image.

File: Makefile [hunk 3]
```diff
@@ -54,6 +61,12 @@ build-image-and-push-linux-amd64: init-buildx
 		--platform linux/amd64 \
 		-t $(REGISTRY)/$(IMG):$(TAG)_linux_amd64 --push . --target debian-base; \
 	}
+	{                                                                   \
+	set -e ;                                                            \
+	docker buildx build \
+		--platform linux/amd64 \
+		-t $(REGISTRY)/$(IMG_NODE_UPDATE_CONTROLLER):$(TAG)_linux_amd64 --push . --target debian-base-node-update-controller; \
+	}
```
Comment (Madhan-SWE) [inline Makefile:69]: No need to build node-update-controller every time when csi driver is built.
Create separate target for building node-update-controller image.

File: Makefile [hunk 4]
```diff
@@ -62,10 +75,18 @@ build-image-and-push-linux-ppc64le: init-buildx
 		--platform linux/ppc64le \
 		-t $(REGISTRY)/$(IMG):$(TAG)_linux_ppc64le --push . --target debian-base; \
 	}
+	{                                                                   \
+	set -e ;                                                            \
+	docker buildx build \
+		--platform linux/ppc64le \
+		-t $(REGISTRY)/$(IMG_NODE_UPDATE_CONTROLLER):$(TAG)_linux_ppc64le --push . --target debian-base-node-update-controller; \
+	}
```
Comment (Madhan-SWE) [inline Makefile:83]: Same as above.

File: Makefile [hunk 5]
```diff
@@ -62,10 +75,18 @@ build-image-and-push-linux-ppc64le: init-buildx
...
 
 build-and-push-multi-arch: build-image-and-push-linux-amd64 build-image-and-push-linux-ppc64le
 	docker manifest create --amend $(REGISTRY)/$(IMG):$(TAG) $(REGISTRY)/$(IMG):$(TAG)_linux_amd64 $(REGISTRY)/$(IMG):$(TAG)_linux_ppc64le
+	docker manifest create --amend $(REGISTRY)/$(IMG_NODE_UPDATE_CONTROLLER):$(TAG) $(REGISTRY)/$(IMG_NODE_UPDATE_CONTROLLER):$(TAG)_linux_amd64 $(REGISTRY)/$(IMG_NODE_UPDATE_CONTROLLER):$(TAG)_linux_ppc64le
```
Comment (Madhan-SWE) [inline Makefile:87]: same as above

File: Makefile [hunk 6]
```diff
@@ -62,10 +75,18 @@ build-image-and-push-linux-ppc64le: init-buildx
...
 	docker manifest push -p $(REGISTRY)/$(IMG):$(TAG)
+	docker manifest push -p $(REGISTRY)/$(IMG_NODE_UPDATE_CONTROLLER):$(TAG)
```
Comment (Madhan-SWE) [inline Makefile:89]: same as above

**Review by Madhan-SWE — CHANGES_REQUESTED** [review]
Initial review

**Comment by Madhan-SWE** [comment]
Having a separate Makefile and Dockerfile placed in ad hoc-controller folder can be a good solution.
@mkumatag what do you think?

**Review by mkumatag — CHANGES_REQUESTED** [review]
As we discussed offline, lets build the binary and include into the same docker image instead of different image.

File: adhoc-controllers/main.go [hunk 7]
```diff
@@ -63,7 +64,13 @@ func main() {
 
 	ctrl.SetLogger(zap.New(zap.UseFlagOptions(&opts)))
 
-	mgr, err := ctrl.NewManager(ctrl.GetConfigOrDie(), ctrl.Options{
```
Comment (mkumatag) [inline adhoc-controllers/main.go:66]: I see `ctrl.GetConfigOrDie` does fallback to in-cluster if kubeconfig supplied, I think we can stick to this function

explanation about this function:
https://github.com/kubernetes-sigs/controller-runtime/blob/e42a8c17945360813e64bde2c4ff742515b750ba/alias.go#L73:L78

**Review by mkumatag — CHANGES_REQUESTED** [review]

File: adhoc-controllers/main.go (hunk 7 above)
Comment (Prajyot-Parab) [reply adhoc-controllers/main.go]: Fixed.

File: Dockerfile [hunk 8]
```diff
@@ -16,15 +16,18 @@ FROM golang:1.17 AS builder
 WORKDIR /go/src/sigs.k8s.io/ibm-powervs-block-csi-driver
 ADD . .
 RUN make driver
+RUN make node-update-controller
```
Comment (mkumatag) [inline Dockerfile:19]:
```suggestion
RUN make driver node-update-controller
```

File: Dockerfile [hunk 9]
```diff
@@ -16,15 +16,18 @@ FROM golang:1.17 AS builder
...
 RUN clean-install ca-certificates e2fsprogs mount udev util-linux xfsprogs bash multipath-tools sg3-utils
 COPY --from=builder /go/src/sigs.k8s.io/ibm-powervs-block-csi-driver/bin/ibm-powervs-block-csi-driver /ibm-powervs-block-csi-driver
+COPY --from=builder /go/src/sigs.k8s.io/ibm-powervs-block-csi-driver/bin/node-update-controller /node-update-controller
 ENTRYPOINT ["/ibm-powervs-block-csi-driver"]
 
 # centos base image
 FROM quay.io/centos/centos:stream8 AS centos-base
 RUN yum install -y util-linux nfs-utils e2fsprogs xfsprogs ca-certificates && yum clean all && rm -rf /var/cache/yum
 COPY --from=builder /go/src/sigs.k8s.io/ibm-powervs-block-csi-driver/bin/ibm-powervs-block-csi-driver /ibm-powervs-block-csi-driver
+COPY --from=builder /go/src/sigs.k8s.io/ibm-powervs-block-csi-driver/bin/node-update-controller /node-update-controller
```
Comment (mkumatag) [inline Dockerfile:32]:
```suggestion
COPY --from=builder /go/src/sigs.k8s.io/ibm-powervs-block-csi-driver/bin/* /ibm-powervs-block-csi-driver
```

>Note: didn't test this one, confirm once

File: Dockerfile [hunk 10]
```diff
@@ -16,15 +16,18 @@ FROM golang:1.17 AS builder
 WORKDIR /go/src/sigs.k8s.io/ibm-powervs-block-csi-driver
 ADD . .
 RUN make driver
+RUN make node-update-controller
 
 # debian base image
 FROM k8s.gcr.io/build-image/debian-base:v2.1.3 AS debian-base
 RUN clean-install ca-certificates e2fsprogs mount udev util-linux xfsprogs bash multipath-tools sg3-utils
 COPY --from=builder /go/src/sigs.k8s.io/ibm-powervs-block-csi-driver/bin/ibm-powervs-block-csi-driver /ibm-powervs-block-csi-driver
+COPY --from=builder /go/src/sigs.k8s.io/ibm-powervs-block-csi-driver/bin/node-update-controller /node-update-controller
```
Comment (mkumatag) [inline Dockerfile:25]:
```suggestion
COPY --from=builder /go/src/sigs.k8s.io/ibm-powervs-block-csi-driver/bin/* /
```

>Note: didn't test this one, confirm once

**Review by mkumatag — CHANGES_REQUESTED** [review]
minor nits to avoid multiple layers in the docker image.

**Review by mkumatag — APPROVED** [review]
/lgtm
/approve

**Pull request merged**